File: packages/baseline-py/src/baseline_py/baseline/read_baseline.py
"""Read the recorded baseline, refusing an incompatible one."""

import json
from pathlib import Path

from baseline_py.baseline.baseline_file import BASELINE_SCHEMA_VERSION, BaselineFile
from baseline_py.config.config_error import ConfigError


def read_baseline(path: Path) -> BaselineFile:
    """Return the baseline, or an empty one when the file is absent.

    An incompatible schema is a configuration error rather than a silent
    reclassification of every finding as new or known.
    """
    if not path.is_file():
        return BaselineFile()
    try:
        document = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, UnicodeDecodeError, json.JSONDecodeError) as error:
        raise ConfigError(f"{path.name} could not be read: {error}") from error
    version = document.get("schema_version")
    if version != BASELINE_SCHEMA_VERSION:
        raise ConfigError(
            f"baseline schema_version must be {BASELINE_SCHEMA_VERSION}, found {version!r}; "
            "run 'baseline-py baseline update' to rewrite it"
        )
    return BaselineFile(
        entries=frozenset(document.get("entries", ())),
        schema_version=version,
        tool_version=document.get("tool_version", ""),
    )

File: packages/db-quality/src/baseline/readBaseline.ts
import { existsSync, readFileSync } from 'node:fs'
import { join } from 'node:path'

import { BASELINE_FILENAME } from '@/baseline/BASELINE_FILENAME.js'
import type { BaselineFile } from '@/baseline/BaselineFile.js'
import { ConfigError } from '@/config/ConfigError.js'

export const readBaseline = (root: string): BaselineFile => {
  const path = join(root, BASELINE_FILENAME)
  if (!existsSync(path)) {
    throw new ConfigError(
      `${BASELINE_FILENAME} not found; run "codeality-db baseline create"`,
    )
  }
  const document = JSON.parse(
    readFileSync(path, 'utf8'),
  ) as Partial<BaselineFile>
  if (document.schemaVersion !== 1 || !Array.isArray(document.entries)) {
    throw new ConfigError(`${BASELINE_FILENAME} is not a version 1 baseline`)
  }
  return {
    schemaVersion: 1,
    toolVersion: document.toolVersion ?? '',
    entries: [...document.entries].sort(),
  }
}
